### PR TITLE
Feat/lw 7973 cip95 sign tx

### PR DESCRIPTION
**Pull request by mirceahasegan**
# Context

Update signing methods to account for new Conway fields:
- [New certificates](https://github.com/input-output-hk/cardano-ledger/blob/efda9c61ef2b75723ff52ee13ee5a16f20cbdb5f/eras/conway/test-suite/cddl-files/conway.cddl#L269C14-L269C14) that require stake and dRep key witnesses
- Voting procedures that require stake or dRep witnesses based on the [Voter type](https://github.com/input-output-hk/cardano-ledger/blob/efda9c61ef2b75723ff52ee13ee5a16f20cbdb5f/eras/conway/test-suite/cddl-files/conway.cddl#L121)

# Proposed Solution
- Refactor witnesses check and foreign key detection in `ownSignaturePaths`, and update `requiresForeignSignatures` to use the same logic.
- Add the logic to check for the new certificates and voting procedures.
- The following certificates cannot be signed by the wallet: `CertificateType.AuthorizeCommitteeHot`, `CertificateType.ResignCommitteeCold`
- [Voting procedures](https://github.com/input-output-hk/cardano-ledger/blob/efda9c61ef2b75723ff52ee13ee5a16f20cbdb5f/eras/conway/test-suite/cddl-files/conway.cddl#L121) that the wallet cannot sign:
  - Voter: `Constitutional Committee Hot KeyHash: 0`
  - Voter: `Constitutional Committee Hot ScriptHash: 1`
  - Voter: `DRep ScriptHash: 3`
- TBD: can we sign voting procedure with `Voter: StakingPool KeyHash: 4`? If it requires stake pool hot key, we can't. At the moment the implementation uses the stake key to sign it.

# TODO in another PR:
- e2e tests. They require serialization/deserialization to be available
- [ ] wait for [serialization work](https://github.com/input-output-hk/cardano-js-sdk/pull/929) to be merged, then update this implementation

### CODE REVIEW

File: packages/key-management/src/util/ownSignatureKeyPaths.ts [hunk 1]
```diff
@@ -185,6 +195,64 @@ const getStakeCredentialKeyPaths = (groupedAddresses: GroupedAddress[], txBody:
...
+      case Cardano.VoterType.dRepKeyHash: {
+        if (voter.credential.hash === Crypto.Hash28ByteBase16.fromEd25519KeyHashHex(dRepKeyHash)) {
+          signatureCheck.derivationPaths.push(DREP_KEY_DERIVATION_PATH);
+        } else {
+          signatureCheck.requiresForeignSignatures = true;
+        }
+        break;
+      }
+      case Cardano.VoterType.stakePoolKeyHash: {
+        // REVIEW: This might require the stake pool hot key
```
Comment (mirceahasegan) [inline packages/key-management/src/util/ownSignatureKeyPaths.ts:233]: Please Review

File: packages/key-management/src/util/stubSignTransaction.ts [hunk 2]
```diff
@@ -9,19 +9,31 @@ import uniqWith from 'lodash/uniqWith';
...
-): Promise<Cardano.Signatures> => {
+export interface StubSignTransactionProps {
+  txBody: Cardano.TxBody;
+  knownAddresses: GroupedAddress[];
+  inputResolver: Cardano.InputResolver;
+  extraSigners?: TransactionSigner[];
+  dRepPublicKey: Crypto.Ed25519PublicKeyHex;
+  signTransactionOptions?: SignTransactionOptions;
+}
+// { additionalKeyPaths = [] }: SignTransactionOptions = {}
```
Comment (mkazlauskas) [inline packages/key-management/src/util/stubSignTransaction.ts:20]: Is this trash?

File: packages/key-management/src/util/ownSignatureKeyPaths.ts [hunk 3]
```diff
@@ -108,7 +323,8 @@ const getRequiredSignersKeyPaths = (
 export const ownSignatureKeyPaths = async (
   txBody: Cardano.TxBody,
   knownAddresses: GroupedAddress[],
-  inputResolver: Cardano.InputResolver
+  inputResolver: Cardano.InputResolver,
+  dRepKeyHash: Crypto.Ed25519KeyHashHex
```
Comment (mkazlauskas) [inline packages/key-management/src/util/ownSignatureKeyPaths.ts:327]: Could just make this optional to not have to use `''` where this is not relevant

**Review by mkazlauskas — APPROVED** [review]
Great work! Seems like we need to commit some change to yarn-project.nix for CI to pass

File: packages/key-management/src/util/ownSignatureKeyPaths.ts (hunk 3 above)
Comment (mirceahasegan) [reply packages/key-management/src/util/ownSignatureKeyPaths.ts]: Done here: [527cc7a](https://github.com/input-output-hk/cardano-js-sdk/pull/919/commits/527cc7a965a4094b538d270c276d0b7ad960365c)

File: packages/key-management/src/util/stubSignTransaction.ts (hunk 2 above)
Comment (mirceahasegan) [reply packages/key-management/src/util/stubSignTransaction.ts]: done: [a1b9746](https://github.com/input-output-hk/cardano-js-sdk/pull/919/commits/a1b9746d3ba8ca2f2e9c5bbd298ba538c1848364)

File: packages/key-management/src/util/ownSignatureKeyPaths.ts (hunk 1 above)
Comment (mirceahasegan) [reply packages/key-management/src/util/ownSignatureKeyPaths.ts]: [a1b9746](https://github.com/input-output-hk/cardano-js-sdk/pull/919/commits/a1b9746d3ba8ca2f2e9c5bbd298ba538c1848364)

File: packages/key-management/src/util/ownSignatureKeyPaths.ts (hunk 3 above)
Comment (rhyslbw) [reply packages/key-management/src/util/ownSignatureKeyPaths.ts]: @mkazlauskas

**Review by mkazlauskas — DISMISSED** [review]
Great work! :rocket: I think we can tackle https://github.com/input-output-hk/cardano-js-sdk/pull/919#discussion_r1330251806 separately

**Review by rhyslbw — DISMISSED** [review]
Great work @mirceahasegan 🦾

**Review by rhyslbw — APPROVED** [review]

File: packages/key-management/src/util/ownSignatureKeyPaths.ts [hunk 4]
```diff
@@ -241,11 +339,15 @@ export const ownSignatureKeyPaths = async (
     ).filter(isNotNil)
   ).map(({ type, index }) => ({ index, role: Number(type) }));
 
+  // TODO: add `proposal_procedure` witnesses.
```
[the remark this answers is not in the document]
Comment (rhyslbw) [reply packages/key-management/src/util/ownSignatureKeyPaths.ts:349]: Has this been captured @mirceahasegan ?

**Pull request merged**